### PR TITLE
Feature/active recording

### PR DESCRIPTION
**Pull request by zaripych**

### CODE REVIEW

File: config/webpack.config.dev.js [hunk 1]
```diff
@@ -181,21 +171,21 @@ module.exports = {
...
       // It's important to do this before Babel processes the JS.
-      {
-        test: /\.(js|mjs|jsx)$/,
-        enforce: 'pre',
-        use: [
-          {
-            options: {
-              formatter: require.resolve('react-dev-utils/eslintFormatter'),
-              eslintPath: require.resolve('eslint'),
+      // {
```
Comment (zaripych) [inline config/webpack.config.dev.js]: I disabled eslinting.

File: config/webpack.config.dev.js [hunk 2]
```diff
@@ -402,24 +388,25 @@ module.exports = {
...
-      formatter: typescriptFormatter,
-    }),
   ].filter(Boolean),
 
   // Some libraries import Node modules but don't use them in the browser.
   // Tell Webpack to provide empty mocks for them so importing them works.
   node: {
     dgram: 'empty',
     fs: 'empty',
+    module: 'empty',
```
Comment (zaripych) [inline config/webpack.config.dev.js]: And disabled the 'module' not found warning. The rest is prettier's involvement.

File: config/webpack.config.dev.js [hunk 3]
```diff
@@ -84,21 +84,17 @@ module.exports = {
   // These are the "entry points" to our application.
```
Comment (zaripych) [inline config/webpack.config.dev.js]: Changes to this file documented below, it's prettier that changed some indentations.

File: src/shared/redux/apps/actions/index.ts [hunk 4]
```diff
@@ -20,9 +20,11 @@ export function errorInfo(error: any): IError {
   } else if (error) {
     const anyError = error as any;
 
+    const name = anyError.status || anyError.code || error.name || 'Error';
+
     return {
-      name: anyError.status || anyError.code || error.name || 'Error',
-      message: error.message || `${error}`,
+      name,
+      message: (error.message !== name && error.message) || 'An unexpected error occurred',
```
Comment (zaripych) [inline src/shared/redux/apps/actions/index.ts]: Some `message` fields from `firebase` are really weird. E.g. just `internal`. Trying to make errors look better :)

File: src/shared/redux/voice-to-text/epics/listenEpic.ts [hunk 5]
```diff
@@ -29,8 +39,16 @@ export const listenEpic: EpicWithState<{}, ReturnType<typeof defaultDeps>> = (ac
...
             action$.ofType<IStopAction>(ActionTypes.STOP),
-            // stop recording when server sent an unrecoverable error:
-            action$.ofType<IErrorAction>(ActionTypes.ERROR),
+            // stop recording after five seconds when server sent an error:
+            action$.ofType<IErrorAction>(ActionTypes.ERROR).pipe(
+              mergeMap(() =>
+                action$.ofType<IResultAction>(ActionTypes.RESULT).pipe(
+                  // if at least one emitted, then finish:
+                  mergeMapTo(empty()),
+                  timeoutWith(5000, of('stop'))
```
Comment (zaripych) [inline src/shared/redux/voice-to-text/epics/listenEpic.ts]: When error happens, do not immediately stop. This is for testing.

File: src/shared/redux/voice-to-text/reducers/listening.ts [hunk 6]
```diff
@@ -76,7 +76,6 @@ export const startStopListeningReducer = (
     const errorAction = action as IErrorAction;
     return {
       ...state,
-      status: 'failed',
```
Comment (zaripych) [inline src/shared/redux/voice-to-text/reducers/listening.ts]: When error happens, do not immediately stop.

File: src/shared/api/audio/startActiveRecording.ts [hunk 7]
```diff
@@ -0,0 +1,165 @@
...
+
+function seekBackInBuffersArr(
+  buffers: BuffType[],
+  timeShift: number,
+  samplesPerBuffer: number,
+  sampleRate: number,
+  bytesPerSample: number
+): BuffType[] {
+  const totalSamplesAvailable = buffers.length * samplesPerBuffer;
+  const targetSampleFromEnd = Math.floor(timeShift * sampleRate);
```
Comment (jaycosaur) [inline src/shared/api/audio/startActiveRecording.ts]: For consistency here just use either Math.floor or Math.trunc as values are always > 0.

File: src/shared/api/audio/startActiveRecording.ts [hunk 8]
```diff
@@ -0,0 +1,165 @@
...
+  sampleRate: number,
+  bytesPerSample: number
+) {
+  const connected = bufferedAudio.pipe(
+    //
+    publish()
+  ) as ConnectableObservable<BuffType[]>;
+  const result = concat(
+    // seek back a little
+    connected.pipe(
```
Comment (jaycosaur) [inline src/shared/api/audio/startActiveRecording.ts]: This is awesome use of RxJS

File: src/shared/redux/voice-to-text/epics/listenEpic.ts (hunk 5 above)
Comment (jaycosaur) [reply src/shared/redux/voice-to-text/epics/listenEpic.ts]: Looks good.

**Pull request merged**